projects/newprofile_ping_nightly_validation.kp/knowledge.md

```diff
@@ -0,0 +1,518 @@
+    .where(appBuildId=lambda x: "20170531" <= x < "20170607") \
+    .records(sc, sample=1.0)
+```
+    fetching 5.53331MB in 1438 files...
+
+
+
+```python
+ping_count = pings.count()
+ping_count
+```
+
+
+
+
+    3840
+
+
+
+### How many pings were sent in-session and how many at shutdown?
+The `new-profile` ping can be sent either during the browsing session, 30 minutes after the browser starts, or at shutdown ([docs](https://gecko.readthedocs.io/en/latest/toolkit/components/telemetry/telemetry/data/new-profile-ping.html#payload-reason)). Let's see how many pings we get in each case.
+
+
+```python
+raw_subset = get_pings_properties(pings, ["id",
+                                          "meta/creationTimestamp",
+                                          "meta/Date",
+                                          "meta/Timestamp",
+                                          "meta/X-PingSender-Version",
+                                          "clientId",
+                                          "environment/profile/creationDate",
+                                          "payload/reason"])
+```
+Discard and count any ping that's missing creationTimestamp or Timestamp.
+
+
+```python
+def pct(a, b):
+    return 100.0 * a / b
+
+subset = raw_subset.filter(lambda p: p["meta/creationTimestamp"] is not None and p["meta/Timestamp"] is not None)
+print("'new-profile' pings with missing timestamps:\t{:.2f}%".format(pct(ping_count - subset.count(), ping_count)))
+```
+    'new-profile' pings with missing timestamps:	0.00%
+
+
+
+```python
+reason_counts = subset.map(lambda p: p.get("payload/reason")).countByValue()
+
+for reason, count in reason_counts.iteritems():
+    print("'new-profile' pings with reason '{}':\t{:.2f}%".format(reason, pct(count, ping_count)))
+```
+    'new-profile' pings with reason 'startup':	24.32%
+    'new-profile' pings with reason 'shutdown':	75.68%
+
+
+This means that, among all the `new-profile` pings, the majority was sent at shutdown. This could mean different things:
+
+- the browsing session lasted less than 30 minutes;
+- we're receiving duplicate pings at shutdown.
+
+### Let's check how many duplicates we've seen
+
+
+```python
+def dedupe(pings, duping_key):
+    return pings\
+            .map(lambda p: (p[duping_key], p))\
+            .reduceByKey(lambda a, b: a if a["meta/Timestamp"] < b["meta/Timestamp"] else b)\
+            .map(lambda pair: pair[1])
+
+deduped_docid = dedupe(subset, "id")
+deduped_docid_count = deduped_docid.count()
+total_duplicates = ping_count - deduped_docid_count
+print("Duplicate pings percentage (by document id): {:.2f}%".format(pct(total_duplicates, ping_count)))
+```
+    Duplicate pings percentage (by document id): 0.21%
+
+
+The 0.21% of ping duplicates is nice, compared to ~1% we usually get from the `main` and `crash` pings. However, nowdays we're running de-duplication by document id at the pipeline ingestion, so this might be a bit higher.
```

Comment:
Note that we have stats on duplicate documents that you can check - there is a parquet table called `telemetry_duplicates_parquet` available in re:dash. I did a quick check and it doesn't look like the duplicate detection flagged any duplicate new-profile pings from June 1 until today:
```
SELECT count(*)
FROM telemetry_duplicates_parquet
WHERE submission_date_s3 >= '20170601'
  AND fields.doc_type = 'new-profile'
```
returns 0 rows